### ISSUE TITLE
Trie: verifyProof() error on nested embeded nodes / batch() function UX issues

### ISSUE THREAD

**Issue by Lion17** [ethereumjs/ethereumjs-monorepo#1055]
When you try to call the verifyProof function for a node that is embedded in another node, which in turn is also embedded, an error occurs.

To reproduce:

```
var Trie = require('merkle-patricia-tree')

trie = new Trie()
var ops = [
   { type: 'put', key: 'dog', value: 'puppy' }
 , { type: 'put', key: 'doge', value: 'coin' }
]

trie.batch(ops,
 () => {
  Trie.prove(trie, 'doge', function (err, prove) {
   if (err) console.log(err)
   else
    Trie.verifyProof(trie.root, 'doge', prove, function (err, value) {
     if (err) console.log(err)
     else console.log(value.toString())
    })
  })
 }
)
```

Output:

> node proof.js
> Error: Unexpected end of proof
>     at Object.exports.verifyProof (/home/local/tmp/node_modules/merkle-patricia-tree/proof.js:122:6)
>     at Function.verifyProof (/home/local/tmp/node_modules/merkle-patricia-tree/index.js:56:32)
>     at /home/local/tmp/proof.js:14:10
>     at /home/local/tmp/node_modules/merkle-patricia-tree/proof.js:33:5
>     at /home/local/tmp/node_modules/merkle-patricia-tree/baseTrie.js:520:18
>     at /home/local/tmp/node_modules/async/dist/async.js:473:16
>     at iteratorCallback (/home/local/tmp/node_modules/async/dist/async.js:1064:13)
>     at /home/local/tmp/node_modules/async/dist/async.js:969:16
>     at Object._return [as return] (/home/local/tmp/node_modules/merkle-patricia-tree/baseTrie.js:545:13)
>     at processNode (/home/local/tmp/node_modules/merkle-patricia-tree/baseTrie.js:325:34)

The reason is that only the variant with one level of embedded nodes is considered. This can be seen in file proof.js:82, where the key length is compared with 1, while with multiple nesting of nodes the key will be longer. In our case, it is equal to 2.

**Comment by holgerd77** [comment]
@ryanio do you have got some insight if this might still be an issue? 🤔

**Comment by ryanio** [comment]
I am not familiar with this issue yet but I will give the example scripts a try to see what could be going on and will report back.

I'll also take a look at the `BatchDBOp[]` type issue, I'm surprised that ts was complaining about the string `type` as it seems to be correct, I'll see if there's something we can do to ease development when trying to use `trie.batch(opts)` without needing any fancy casting.

**Comment by zmitton** [comment]
The written error reported is "unexpected end of proof", and its coming from the proof.js file. Both that error meseage and the file have been eliminated.  I believe I wrote a similar situation to return "missing node error". 

But now that I look through the code, I cant find that error anymore. I think it should be somewhere around here: 

https://github.com/ethereumjs/ethereumjs-monorepo/blob/2357981d89b865361126e2fa3e77f90780a8f3ef/packages/trie/src/baseTrie.ts#L268

Note: Although the behavior of the tree may be to return `null` when a missing key is looked up, the behavior of the *underlaying db* (not part of the API!) should return an error when a missing key is looked up. Two completely different things!

something like 

```javascript
    if (value) {
      foundNode = decodeNode(value)
    } else{
      return new Error("missing trie node")
    }
```

I could be wrong. I dont have time to work on this but I hope the information is helpful

**Comment by Chynia828** [comment]
#1055

**Comment by acolytec3** [comment]
I can't speak to t

> The written error reported is "unexpected end of proof", and its coming from the proof.js file. Both that error meseage and the file have been eliminated. I believe I wrote a similar situation to return "missing node error".
> 
> But now that I look through the code, I cant find that error anymore. I think it should be somewhere around here:
> 
> https://github.com/ethereumjs/ethereumjs-monorepo/blob/2357981d89b865361126e2fa3e77f90780a8f3ef/packages/trie/src/baseTrie.ts#L268
> 
> Note: Although the behavior of the tree may be to return `null` when a missing key is looked up, the behavior of the _underlaying db_ (not part of the API!) should return an error when a missing key is looked up. Two completely different things!
> 
> something like
> 
> ```js
>     if (value) {
>       foundNode = decodeNode(value)
>     } else{
>       return new Error("missing trie node")
>     }
> ```
> 
> I could be wrong. I dont have time to work on this but I hope the information is helpful

I don't know that I fully understand the logic of why the db not finding a node should generate an error versus a null value but as [here[(https://github.com/ethereumjs/ethereumjs-monorepo/blob/008a4c3a044908614d2f0da22583537b7b383289/packages/trie/src/db.ts#L44) our code does check for the key not found error condition in the db and then just returns null to the trie module in such cases.  It throws on other errors.

**Comment by holgerd77** [comment]
Is this still an issue or can we close after 3 years? 😅

Lol. 😛

**Comment by ryanio** [comment]
I think this is okay now, we might have solved it as part of #1368, I ran the example script in https://github.com/ethereumjs/ethereumjs-monorepo/issues/1055#issuecomment-740542039 and got back a valid value: `<Buffer 63 6f 69 6e>`

If someone would like to follow up please open a new issue with a reproducible script so we can help.
